### PR TITLE
Adjust size-speed trade-offs to save stack space

### PR DESCRIPTION
**Pull request by c1728p9**
Both the values MBEDTLS_MPI_MAX_SIZE and MBEDTLS_MPI_WINDOW_SIZE
declare buffers on the stack. The default values of
MBEDTLS_MPI_MAX_SIZE=1024 and MBEDTLS_MPI_WINDOW_SIZE=6 cause the
tls-client example to just barely overflow the 4K main stack in some
configurations. This patch sets both of those defines to smaller
values to decrease stack size.

When building ARMCC with the default profile this reduces stack usage
of the tls-client example from ~4184 to ~2372 bytes.

### CODE REVIEW

**Comment by c1728p9** [comment]
@andresag01 @sbutcher-arm please feel free to adjust this patch as you see fit. I don't know the full impact of these settings aside from the fact that they can be used to trade speed for memory usage.

File: tls-client/mbedtls_entropy_config.h [hunk 1]
```diff
@@ -34,3 +34,5 @@
 #endif /* !MBEDTLS_ENTROPY_HARDWARE_ALT && !MBEDTLS_ENTROPY_NV_SEED */
 
 #define MBEDTLS_SHA1_C
+#define MBEDTLS_MPI_MAX_SIZE        256
+#define MBEDTLS_MPI_WINDOW_SIZE     3
```
Comment (c1728p9) [inline tls-client/mbedtls_entropy_config.h]: Would it be better to put this in the mbed-os default tls config rather than (or in addition to) here?

**Comment by yanesca** [comment]
Hi @c1728p9, thank you for the PR! Although this seems to work, theoretically it shouldn't: there is some bug or unforeseen behaviour that makes it work. Therefore we would like to rework it. Thanks again!

**Comment by yanesca** [comment]
We have worked out the problem in the background and now we can use the `MBEDTLS_MPI_MAX_SIZE` you provided safely.

**Comment by kjbracey-arm** [comment]
FAO @yogpan01, @SeppoTakalo, @hasnainvirk - probably worth clamping down on these config params for basically all 6LoWPAN stuff.

**Comment by yogpan01** [comment]
@TeroJaasko FYI

**Comment by kjbracey-arm** [comment]
@TeroJaasko - I think for most 6LoWPAN setups, MPI size could probably be 32, as I believe we only need to do ECC with 256-bit keys (secp256r1 curve).

Of course, really there should be a proper mbedTLS JSON configuration mechanism allowing modules to collate their requirements, but in the absense of that you could adjust it for any apps (somehow) knowing that no component in the system is using any crypto with longer keys.

**Comment by sbutcher-arm** [comment]
@kjbracey-arm @TeroJaasko @c1728p9 c1728p9 - You can't change the `MBEDTLS_MPI_MAX_SIZE` without considering other options, otherwise you will get unexpected side effects, and when you do change these configurations, you don't just change it for your application use but for the whole platform as mbed TLS is a common resource.

I do not recommend deviating from the mbed OS defaults for normal usage, unless you're really confident in what you're doing and you appreciate we, the mbed TLS team, are unlikely to support you.

**Comment by c1728p9** [comment]
@sbutcher-arm is changing MBEDTLS_MPI_MAX_SIZE to 256 in this PR acceptable? This buffer declared by MBEDTLS_MPI_MAX_SIZE is on the stack, and if this change is reverted then stack overflows will start occurring again.

**Comment by kjbracey-arm** [comment]
Unfortunately we have no proper configuration mechanism for the system - there is no way of combining the requirements via JSON. There's no way for the current Wi-Fi driver to say "I need crypto X,Y and Z and keys of at least this length" and combine that with CoAP saying "I need ECC and 256-bit keys" and with the 802.15.4 saying "I need AES".

So it's totally up to the application to control mbed TLS, using its godlike knowledge of every part of the system. There's no current alternative. (Apart from just not fitting in the system at all).

We would greatly appreciate the mbed TLS team offering support by offering a proper system configuration mechanism so it wasn't necessary for apps to do anything like this.

**Comment by sbutcher-arm** [comment]
@c1728p9 - We sat down and determined that this PR was acceptable based on the configured key sizes. The problem comes because this is a tweak, it's very possible to configure a value that is invalid and incompatible with other configurations because the build doesn't do sanity checks. eg. You could set a low `MBEDTLS_MPI_MAX_SIZE` value, yet also configure key sizes that would require a much larger value - it could all compile and even run until you run out of space for larger key sizes.

My point was really aimed at @kjbracey-arm's comment:
> MPI size could probably be 32, as I believe we only need to do ECC with 256-bit keys (secp256r1 curve)

@kjbracey-arm - There's two issues here - one about the process of configuring mbed TLS and another concerning configuration of the platform. I'll take those up by email as they need wider discussion with other people.

**Pull request merged**